### PR TITLE
Add methods to PrefixIterator to support iterating from a specific key

### LINKED ISSUE #9213: Iterate one storage map over blocks
We have a storage map with huge amount of keys. Is it possible to iterate one storage map over blocks? I cannot store the PrefixIterator into the DB since it cannot be encoded. Meanwhile, fields of PrefixIterator are private. I cannot build a new PrefixIterator with a given previous_key.

### PR DESCRIPTION
**Pull request by KiChjang**
Fixes #9213.

This adds methods that aid in paginating storage iterators. I imagine that this will be used heavily in storage migrations that span multiple blocks, or other sorts of consensus mechanisms that span multiple blocks and require interaction with storage.

### CODE REVIEW

File: frame/support/src/storage/mod.rs [hunk 1]
```diff
@@ -427,6 +460,36 @@ pub trait IterableStorageNMap<K: ReversibleKeyGenerator, V: FullCodec>: StorageN
...
+	fn partial_key_value_decode_fn<KP>(
+		raw_key_without_prefix: &[u8],
+		raw_value: &[u8],
+	) -> Result<(<K as HasKeyPrefix<KP>>::Suffix, V), codec::Error>
+	where K: HasReversibleKeyPrefix<KP>;
+
+	/// The decode function used while iterating through the prefix keys in
+	/// [`IterableStorageNMap::iter_key_prefix`].
+	///
+	/// A type parameter denoting the type of the key prefix must be provided. See
```
Comment (KiChjang) [inline frame/support/src/storage/mod.rs:476]: This documentation is not great, I'll need to rephrase this to properly explain what `KP` does.

File: frame/support/src/storage/mod.rs [hunk 2]
```diff
@@ -781,6 +875,36 @@ pub struct KeyPrefixIterator<T> {
...
+	) -> Self {
+		KeyPrefixIterator {
+			prefix,
+			previous_key,
+			drain: false,
+			closure: decode_fn,
+		}
+	}
+
+	/// Get the next key that is about to be iterated upon and return it.
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:898]: I think the doc is confusing, this returns the last key that has been iterated upon or to start iterating after.
I.e. the key to continue iterating after.

File: frame/support/src/storage/mod.rs [hunk 3]
```diff
@@ -781,6 +875,36 @@ pub struct KeyPrefixIterator<T> {
...
+	///
+	/// A `decode_fn` function must also be supplied, and it takes in a `&[u8]` parameter, returning
+	/// a `Result` containing the decoded key type `T` if successful, and a `codec::Error` on
+	/// failure. The `&[u8]` argument represents the raw, undecoded key without the prefix of the
+	/// current item.
+	pub fn new(
+		prefix: Vec<u8>,
+		previous_key: Vec<u8>,
+		decode_fn: fn(&[u8]) -> Result<T, codec::Error>,
+	) -> Self {
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:889]: I feel like better than exposing all the functions `partial_key_decode` ... we can do a better API.
Currently it is a bit error prone considering user have to create iterator with the correct decode_fn.
Instead maybe we can create a function: `start_iterating_after` (or a better name) which would set the `previous_key` to iterate after.
Then people can just do `MyStorageMap::iter().start_iterating_after(the old raw key)`
Or maybe even better `MyStorageMap::iter_after(my_key_not_raw)`.

what do you think ?

Comment (thiolliere) [reply frame/support/src/storage/mod.rs:889]: I feel like just adding a function `set_previous_key` is more handy as you don't need to create a PrefixIterator and give the correct decode_fn.
Anyhow we can keep the `

File: frame/support/src/storage/mod.rs [hunk 4]
```diff
@@ -1413,6 +1537,59 @@ mod test {
 		});
 	}
 
+	#[test]
+	fn prefix_iterator_pagination_works() {
+		TestExternalities::default().execute_with(|| {
+			use crate::storage::generator::StorageMap;
+			use crate::hash::Twox64Concat;
+			struct MyStorageMap;
```
Comment (kianenigma) [inline frame/support/src/storage/mod.rs:1545]: `generate_storage_alias` might help here.

File: frame/support/src/storage/mod.rs (hunk 3 above)
Comment (kianenigma) [reply frame/support/src/storage/mod.rs:889]: I totally agree with @thiolliere here. My expectation of the API was that the user stores a key, and next to `StorageMap::iter()`, they can do `StorageMap::iter_from(stored_key)`.

**Review by kianenigma — COMMENTED** [review]
I feel like the PR is correct, but incomplete. I don't see any change in public interfaces like `IterableStorageMap` etc. My assumption was that there's be a new `fn iter_from()` added to these traits, which would create an iterating type from `KeyPrefixIterator` or similar types.

File: frame/support/src/storage/mod.rs (hunk 3 above)
Comment (KiChjang) [reply frame/support/src/storage/mod.rs:889]: Yeah, this is definitely a better idea, I just couldn't figure out how to create a brand new `PrefixIterator` and start iterating from a given key since it required a decode_fn.

File: frame/support/src/storage/mod.rs [hunk 5]
```diff
@@ -321,10 +321,18 @@ pub trait IterableStorageMap<K: FullEncode, V: FullCodec>: StorageMap<K, V> {
 	/// this, you'll get undefined results.
 	fn iter() -> Self::Iterator;
 
+	/// Enumerate all elements in the map starting from a specified `starting_key` in no particular
+	/// order. If you alter the map while doing this, you'll get undefined results.
+	fn iter_from(starting_key: Vec<u8>) -> Self::Iterator;
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:326]: The doc should precise that it iterates after the starting_key, the iterator will never return the (key/value) pair corresponding to the starting_key.

same for other methods

Also I think we should be more precise on the doc for iterator, this doesn't iterate in no particular order, it iterates on the lexicographic order of the encoded key. This what allows us to resume iterator

Also maybe it is better to name it `starting_raw_key` to convey the sense that this is the final key: encoded key with storage prefixes.

File: frame/support/src/storage/mod.rs [hunk 6]
```diff
@@ -716,6 +774,42 @@ pub struct PrefixIterator<T> {
...
+		PrefixIterator {
+			prefix,
+			previous_key: next_key,
+			drain: false,
+			closure: decode_fn,
+		}
+	}
+
+	/// Get the last key that has been iterated upon and return it.
+	pub fn next_key(&self) -> &[u8] {
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:799]: i still feel weird naming this `next_key` when it is actually the `last_key`.

I would personally name it `last_raw_key`, or `current_raw_key`.

File: frame/support/src/storage/mod.rs [hunk 7]
```diff
@@ -716,6 +774,42 @@ pub struct PrefixIterator<T> {
...
+		&self.previous_key
+	}
+
+	/// Get the prefix that is being iterated upon for this iterator and return it.
+	pub fn prefix(&self) -> &[u8] {
+		&self.prefix
+	}
+
+	/// Set the next key that the iterator should start iterating upon.
+	pub fn set_next_key(&mut self, next_key: Vec<u8>) {
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:809]: same I feel like `next_key` is a confusing name, `set_last_raw_key` or `set_current_raw_key` is a better name IMO

File: frame/support/src/storage/mod.rs (hunk 5 above)
Comment (KiChjang) [reply frame/support/src/storage/mod.rs:326]: Is it 100% accurate that it iterates on the lexicographical order? Because I tried calling `next()` on an iterator in the test, and the element that gets returned seemed to be random. Concretely speaking, I had a `MyMap` that has numbered keys from 1 to 10, calling `next()` on the iterator yielded the `6` key, which does not look to be lexicographically ordered at all.

Comment (thiolliere) [reply frame/support/src/storage/mod.rs:326]: it should be the lexicographic order of the encoded key.
So for the keys represented by array of byte: `[1, 255, 3, 1]`, `[1, 255]`, `[1, 255, 4, 0]`, the order must be
 `[1, 255]`, `[1, 255, 3, 1]`, `[1, 255, 4, 0]`.
 
the doc for next key say so https://substrate.dev/rustdocs/latest/sp_io/storage/fn.next_key.html

Comment (thiolliere) [reply frame/support/src/storage/mod.rs:326]: > Concretely speaking, I had a MyMap that has numbered keys from 1 to 10, calling next() on the iterator yielded the 6 key, which does not look to be lexicographically ordered at all.

It is because the hashes don't follow the same order as the key.

File: frame/support/src/storage/generator/double_map.rs [hunk 8]
```diff
@@ -358,6 +358,15 @@ impl<
...
 	}
 
+	fn iter_prefix_from(
+		k1: impl EncodeLike<K1>,
+		starting_raw_key: Vec<u8>,
+	) -> Self::PrefixIterator {
+		let mut iter = Self::iter_prefix(k1);
+		iter.set_last_raw_key(starting_raw_key);
+		iter
+	}
```
Comment (thiolliere) [inline frame/support/src/storage/generator/double_map.rs:368]: I think every new method implementation should have a quick test

File: frame/support/src/storage/mod.rs [hunk 9]
```diff
@@ -321,10 +321,18 @@ pub trait IterableStorageMap<K: FullEncode, V: FullCodec>: StorageMap<K, V> {
 	/// this, you'll get undefined results.
 	fn iter() -> Self::Iterator;
 
+	/// Enumerate all elements in the map after a specified `starting_raw_key` in no particular
+	/// order. If you alter the map while doing this, you'll get undefined results.
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:325]: I still think we need to say what is the order. If there is no specified order, how can people do some partial iteration and resume them.

File: frame/support/src/storage/mod.rs [hunk 10]
```diff
@@ -716,6 +774,42 @@ pub struct PrefixIterator<T> {
 }
 
 impl<T> PrefixIterator<T> {
+	/// Creates a new `PrefixIterator`, iterating after `next_key` and filtering out keys that are
```
Comment (thiolliere) [inline frame/support/src/storage/mod.rs:777]: I would personally name it `from_key` or `after_key` or `previous_key` instead of `next_key` because it is not the next fetched key.
And I would write it as a note.

**Review by thiolliere — COMMENTED** [review]
looks good but I would prefer having
* some more tests
* order specified
* some naming nitpicks

File: frame/support/src/storage/mod.rs [hunk 11]
```diff
@@ -1413,6 +1542,52 @@ mod test {
...
+			remaining.sort();
+			assert_eq!(remaining.len(), 8);
+			assert_eq!(remaining, vec![3, 4, 5, 6, 7, 8, 9, 10]);
+
+			final_vec.extend_from_slice(&remaining);
+			final_vec.sort();
+
+			assert_eq!(final_vec, vec![1, 2, 3, 4, 5, 6, 7, 8, 9, 10]);
+		});
+	}
```
Comment (emostov) [inline frame/support/src/storage/mod.rs:1589]: The new doc comments say something along the lines of "Enumerate all elements in the map after a specified `starting_raw_key` in no particular", but the test here makes it look like we enumerate in the order of insertion.

So to clarify:
- If we are to iterate over a map single, non mutated map multiple times, will we get the same ordering each time?
- If we do not get the same ordering, what is the point of having a starting key? Seems like that key could be at any position in the ordering if ordering is variant, which does not seem useful

File: frame/support/src/storage/mod.rs (hunk 5 above)
Comment (emostov) [reply frame/support/src/storage/mod.rs]: > it should be the lexicographic order of the encoded and hashed key. (the final key put in the trie)
So for the keys represented by array of byte: [1, 255, 3, 1], [1, 255], [1, 255, 4, 0], the order must be
[1, 255], [1, 255, 3, 1], [1, 255, 4, 0].

^ not sure what convention is, but imo would be nice to include that in the docs too and a note pointing out

> the hashes don't follow the same order as the key.

For me at least, this bits of info connected the missing pieces to understanding the iteration order

**Review by thiolliere — APPROVED** [review]
looks good to me

File: frame/support/src/storage/generator/double_map.rs [hunk 12]
```diff
@@ -371,17 +380,30 @@ impl<
...
+
 	fn drain_prefix(k1: impl EncodeLike<K1>) -> Self::PrefixIterator {
 		let mut iterator = Self::iter_prefix(k1);
 		iterator.drain = true;
 		iterator
 	}
 
 	fn iter() -> Self::Iterator {
+		Self::iter_from(G::prefix_hash())
```
Comment (kianenigma) [inline frame/support/src/storage/generator/double_map.rs:399]: Wee bit strange here. `iter_from` is already reading `let prefix = G::prefix_hash();`, but we pass in `G::prefix_hash()` to it again. I'd make the `starting_raw_key` an `Option`.

File: frame/support/src/storage/generator/map.rs [hunk 13]
```diff
@@ -144,10 +144,15 @@ impl<
 
 	/// Enumerate all elements in the map.
 	fn iter() -> Self::Iterator {
+		Self::iter_from(G::prefix_hash())
```
Comment (kianenigma) [inline frame/support/src/storage/generator/map.rs:147]: same grumble.

**Review by kianenigma — APPROVED** [review]
I've got one minor note, otherwise looks good now.

File: frame/support/src/storage/generator/double_map.rs (hunk 12 above)
Comment (KiChjang) [reply frame/support/src/storage/generator/double_map.rs:399]: I guess we can just revert this and implement `iter_from` in terms of `iter` by calling `set_previous_key` on the resulting `PrefixIterator`.

**Comment by KiChjang** [comment]
bot merge
